Add additional LayC guidance in docs (#3193)
SonarSource/rspec: docs/description.adoc

Reviewer: Another thing: The PR title does not reflect the intention of the PR

Author: Good point. I have made some updates.

See also the <<styling_guide.adoc#,Styling Guidelines>>, <<tone_guide.adoc#,Tone Guide>> and https://docs.sonarqube.org/latest/extension-guide/adding-coding-rules/#coding-rule-guidelines[Coding rule guidelines].

== General guidance

Most sections and subsections of the educational rule format are optional, only the `Why is this an issue?` main section is mandatory. 
The goal is to provide the right level of guidance so that users get the right information from the rules at the right time.
Do not feel obliged to use every section or sub-section if omitting them would lead to a better user experience.

=== Guidelines if you aren’t sure where something belongs in a rule
* If you need the information to fix the issue, but only the first time, it probably belongs in ‘Why is this an issue?’
* If you need the information to fix the issue every time, it probably belongs in ‘How to fix it?’
* If you don’t need the information to fix the issue, but it will help users grow their knowledge, it probably belongs in ‘Resources’ (this maps to the 'More Info' tab in the products)

By being careful about what goes where, we help to ensure that users get exactly what they need, when they need it.

> 

Author: Agreed, I have moved it as suggested.

=== Guidelines on content focus

We want to help users to create Clean Code. Rule content should focus on the manner in which the code is not clean, why this is an issue, and how to remedy this. 

> Rules should talk about the potential impact on software quality in the 'What is the potential impact?' sub-section.

Author: That's a good point. I have tried to add an example to clarify things a bit.

Third-level and fourth-level titles (`+=== Title t3+` and `+==== Title t4+`) are not checked for this type of rule.

> === 2. Educational Rule Format

Author: Yes, updated.

Start at the basics and go into more depth to explain the concepts behind this type of issue. This is most likely the place where a lot of the content will be. +
This is the place to tell the “story” of the rule, including the impact of leaving it unfixed. We should include code samples wherever needed to make it easier to understand
what is going on. This can be in the form of noncompliant and compliant code in a single code box (noncompliant lines should always be highlighted with the corresponding comment
“// Noncompliant” optionally followed by some explanation) if that is clearer. This first tab could use a blog post style with a free format explaining what the rule is
detecting and why. Feel free to use “What is the potential impact?” title if it makes sense, or any other titles you find useful. +
Goal: Understand the concepts behind an issue and why it matters.
+
** *What is the potential impact?* (level 3 title) [Optional]
+

> What is the risk for me? What can go wrong?

Reviewer: ```suggestion
This subsection aims to address questions such as the following:
* What is the risk for me?
* What is the risk for my organization?
* What is the risk for the users?
* What can go wrong?
```

> This section can include multiple concrete threats as well. This is the place to talk about the impacts to the software(e.g. security, maintainability, etc.)

Reviewer: ```suggestion
This section aims to talk about the impacts on the software, depending on the code qualities this issue is linked to: security, reliability, or maintainability. *(see [docs/metadata#impacts](https://github.com/SonarSource/rspec/blob/master/docs/metadata.adoc#code-field).)*

Consequently, this section can also talk about the impacts on the ecosystem of this software. For example, impacts on the organization, on the users, and impact in terms of regulations.
```

Start at the basics and go into more depth to explain the concepts behind this type of issue. This is most likely the place where a lot of the content will be. +
This is the place to tell the “story” of the rule, including the impact of leaving it unfixed. We should include code samples wherever needed to make it easier to understand
what is going on. This can be in the form of noncompliant and compliant code in a single code box (noncompliant lines should always be highlighted with the corresponding comment
“// Noncompliant” optionally followed by some explanation) if that is clearer. This first tab could use a blog post style with a free format explaining what the rule is
detecting and why. Feel free to use “What is the potential impact?” title if it makes sense, or any other titles you find useful. +
Goal: Understand the concepts behind an issue and why it matters.
+
** *What is the potential impact?* (level 3 title) [Optional]
+
This subsection aims to address questions such as the following: +

> What is the risk for me?  What is the risk for my organization? What is the risk for the users? What can go wrong?

Reviewer: ```suggestion
What is the risk for me?  What is the risk for my organization? What is the risk for the users? What can go wrong? +
```

> This section aims to talk about the impacts on the software, depending on the code qualities this issue is linked to: security, reliability, or maintainability. *(see <<metadata.adoc#,impacts>>.)*

Reviewer: ```suggestion
This section aims to talk about the impacts on the software, depending on the code qualities this issue is linked to: security, reliability, or maintainability. *(see <<metadata.adoc#,code-field>>.)*
```

> This section aims to talk about the impacts on the software, depending on the code qualities this issue is linked to: security, reliability, or maintainability. *(see <<metadata.adoc#,code-field>>.)*

Reviewer: ```suggestion
This section aims to talk about the impacts on the software, depending on the code qualities this issue is linked to: security, reliability, or maintainability. **(see <<metadata.adoc#code-field,Impacts>>.)**
```

> This section aims to talk about the impacts on the software, depending on the code qualities this issue is linked to: security, reliability, or maintainability. *(see [docs/metadata#impacts](https://github.com/SonarSource/rspec/blob/master/docs/metadata.adoc#code-field).)*

Reviewer: ```suggestion
This section aims to talk about the impacts on the software, depending on the code qualities this issue is linked to: security, reliability, or maintainability. *(see <<metadata.adoc#,impacts>>.)*
```

Furthermore, the display name of the framework has to match an allowed framework
display name, as defined in <<header_names/allowed_framework_names.adoc#,this allowed framework names file>>.

==== General guidance

Most sections and subsections of the Learn as You Code rule format are optional, only the `Why is this an issue?` main section is mandatory. 
The goal is to provide the right level of guidance so that users get the right information from the rules at the right time.
Do not feel obliged to use every section or sub-section if omitting them would lead to a better user experience.

> ===== Guidelines if you aren’t sure where something belongs in a rule

Reviewer: ```suggestion
==== Guidelines if you aren’t sure where something belongs in a rule
```

* If you need the information to fix the issue, but only the first time, it probably belongs in ‘Why is this an issue?’
* If you need the information to fix the issue every time, it probably belongs in ‘How to fix it?’
* If you don’t need the information to fix the issue, but it will help users grow their knowledge, it probably belongs in ‘Resources’ (this maps to the 'More Info' tab in the products)

By being careful about what goes where, we help to ensure that users get exactly what they need, when they need it.

> ===== Guidelines on content focus

Reviewer: ```suggestion
==== Guidelines on content focus
```
